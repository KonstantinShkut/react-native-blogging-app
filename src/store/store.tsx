import React, { useEffect } from 'react';
import auth, { FirebaseAuthTypes } from '@react-native-firebase/auth';

interface State {
  isLoggedIn: boolean | null;
  user: FirebaseAuthTypes.User | null;
}

const initialState: State = {
  isLoggedIn: null,
  user: null,
};

const storeContext = React.createContext<any>(null);

const reducer = (state: State, action: any) => {
  switch (action.type) {
    case 'LOGIN':
      return { ...state, isLoggedIn: true, user: action.user };
    case 'LOGOUT':
      return { ...state, isLoggedIn: false, user: null };
    default:
      return state;
  }
};

export const StoreProvider = ({ children }: any) => {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  const actions = React.useMemo(
    () => ({
      authenticate: async (user: FirebaseAuthTypes.User | null = null) => {
        if (!user) {
          const { currentUser } = await auth();
          user = currentUser;
        }
        if (user) {
          dispatch({
            type: 'LOGIN',
            user: user,
          });
        }
      },

      signup: async (email: string, password: string, name: string) => {
        const cred = await auth().createUserWithEmailAndPassword(
          email,
          password,
        );
        await cred.user.updateProfile({
          displayName: name,
        });
        await actions.authenticate(cred.user);
      },

      login: async (email: string, password: string) => {
        const cred = await auth().signInWithEmailAndPassword(email, password);
        await actions.authenticate(cred.user);
      },

      logout: async () => {
        await auth().signOut();
        dispatch({
          type: 'LOGOUT',
        });
      },
    }),
    [],
  );

  useEffect(() => {
    (async () => {
      await actions.authenticate();
    })();
  }, []);

  return (
    <storeContext.Provider value={{ ...state, ...actions }}>
      {children}
    </storeContext.Provider>
  );
};

export const useStore = () => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
