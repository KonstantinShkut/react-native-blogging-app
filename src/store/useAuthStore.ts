import React, { useEffect } from 'react';
import auth, { FirebaseAuthTypes } from '@react-native-firebase/auth';
import { User } from '../types';

const LOGIN = 'LOGIN';
const LOGOUT = 'LOGOUT';

interface State {
  isLoggedIn: boolean | null;
  user: User | null;
}

const initialState: State = {
  isLoggedIn: null,
  user: null,
};

interface LoginAction {
  type: typeof LOGIN;
  user: User;
}

interface LogoutAction {
  type: typeof LOGOUT;
}

type ActionType = LoginAction | LogoutAction;

const reducer = (state: State, action: ActionType) => {
  switch (action.type) {
    case 'LOGIN':
      const { user } = action;
      return { ...state, isLoggedIn: true, user };
    case 'LOGOUT':
      return { ...state, isLoggedIn: false, user: null };
    default:
      return state;
  }
};

export const useAuthStore = () => {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  const actions = React.useMemo(
    () => ({
      authenticate: async (user: FirebaseAuthTypes.User | null = null) => {
        if (!user) {
          const { currentUser } = await auth();
          user = currentUser;
        }
        if (user) {
          dispatch({
            type: 'LOGIN',
            user: {
              id: user.uid,
              name: user.displayName as string,
              email: user.email as string,
            },
          });
        }
      },

      signup: async (email: string, password: string, name: string) => {
        const cred = await auth().createUserWithEmailAndPassword(
          email,
          password,
        );
        await cred.user.updateProfile({
          displayName: name,
        });
        await actions.authenticate(cred.user);
      },

      login: async (email: string, password: string) => {
        const cred = await auth().signInWithEmailAndPassword(email, password);
        await actions.authenticate(cred.user);
      },

      logout: async () => {
        await auth().signOut();
        dispatch({
          type: 'LOGOUT',
        });
      },
    }),
    [],
  );

  useEffect(() => {
    (async () => {
      await actions.authenticate();
    })();
  }, []);

  return { ...state, ...actions };
};
