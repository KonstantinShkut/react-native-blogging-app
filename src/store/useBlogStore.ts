import React, { useEffect, useRef } from 'react';
import {
  addPost,
  deletePost,
  loadPosts,
  subscribeToPosts,
  updatePost,
} from '../services/blog';
import { PostData, PostItem } from '../services/blog/types';

const LOADING = 'LOADING';
const LOADED = 'LOADED';
const LOADED_MORE = 'LOADED_MORE';

interface State {
  isLoading: boolean;
  posts: PostItem[];
}

const initialState: State = {
  isLoading: false,
  posts: [],
};

interface LoadingAction {
  type: typeof LOADING;
}

interface LoadedAction {
  type: typeof LOADED;
  posts: PostItem[];
}

interface LoadedMoreAction {
  type: typeof LOADED_MORE;
  posts: PostItem[];
}

type ActionType = LoadingAction | LoadedAction | LoadedMoreAction;

const reducer = (state: State, action: ActionType) => {
  switch (action.type) {
    case LOADING: {
      return { ...state, isLoading: true };
    }
    case LOADED: {
      const { posts } = action;
      if (state.posts) {
        const loaded = posts.map((post) => {
          const existingPost =
            state.posts &&
            state.posts.find(
              (p) => p.id === post.id && p.updated.isEqual(post.updated),
            );
          return existingPost || post;
        });
        return { ...state, isLoading: false, posts: loaded };
      } else {
        return { ...state, isLoading: false, posts };
      }
    }
    case LOADED_MORE: {
      const { posts } = action;
      return {
        ...state,
        isLoading: false,
        posts: posts.length ? [...state.posts, ...posts] : state.posts,
      };
    }
    default:
      return state;
  }
};

export const useBlogStore = () => {
  const unsubscribe = useRef<() => void>();
  const [state, dispatch] = React.useReducer(reducer, initialState);
  const actions = React.useMemo(
    () => ({
      load: async () => {
        dispatch({ type: LOADING });
        const posts = await loadPosts();
        dispatch({ type: LOADED, posts });
      },
      loadMore: async () => {
        if (state.isLoading || !state.posts.length) {
          return;
        }
        dispatch({ type: LOADING });
        const lastPost = state.posts[state.posts.length - 1];
        const posts = await loadPosts(lastPost.id);
        dispatch({ type: LOADED_MORE, posts });
      },
      subscribe: async () => {
        const callback = (posts: PostItem[]) =>
          dispatch({ type: LOADED, posts });
        unsubscribe.current && unsubscribe.current();
        unsubscribe.current = await subscribeToPosts(callback);
      },
      add: async (data: PostData) => {
        await addPost(data);
        await actions.load();
      },
      delete: async (post: PostItem) => {
        await deletePost(post);
        await actions.load();
      },
      update: async (post: PostItem, data: PostData) => {
        await updatePost(post, data);
        await actions.load();
      },
    }),
    [state],
  );

  useEffect(() => {
    (async () => {
      await actions.load();
      return () => {
        return unsubscribe.current && unsubscribe.current();
      };
    })();
  }, []);

  return { ...state, ...actions };
};
