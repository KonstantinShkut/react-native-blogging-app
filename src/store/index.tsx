import React from 'react';
import { useAuthStore } from './useAuthStore';
import { useBlogStore } from './useBlogStore';

const storeContext = React.createContext<Store | null>(null);

interface Store {
  auth: ReturnType<typeof useAuthStore>;
  blog: ReturnType<typeof useBlogStore>;
}

export const StoreProvider = ({ children }: any) => {
  const store = {
    auth: useAuthStore(),
    blog: useBlogStore(),
  };
  return (
    <storeContext.Provider value={store}>{children}</storeContext.Provider>
  );
};

export const useStore = () => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
