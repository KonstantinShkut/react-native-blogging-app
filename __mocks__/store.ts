// @ts-nocheck
import * as store from '../src/store';

export const mockStore = jest.spyOn(store, 'useStore');
mockStore.mockReturnValue({
  auth: {},
  blog: {},
});
