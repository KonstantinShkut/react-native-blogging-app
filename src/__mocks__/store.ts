// @ts-nocheck
import * as store from '../store';

export const mockStore = jest.spyOn(store, 'useStore');
mockStore.mockReturnValue({
  auth: {},
  blog: {},
});
