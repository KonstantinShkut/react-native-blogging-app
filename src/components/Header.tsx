import React from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { colorPrimary, colorSecondary, header } from '../styles';
import { useStore } from '../store';
import { Icon } from 'react-native-elements';

interface Props {
  // prop: string;
}

export const Header = (props: Props) => {
  const navigation = useNavigation();
  const { auth } = useStore();
  const { user } = auth;

  const onPressProfile = () => navigation.navigate('Profile');
  // const onPressProfile = () => navigation.navigate('Logout');

  return (
    <View style={styles.container}>
      <View style={styles.row}>
        <View style={styles.left}>
          <Text style={header}>All Posts</Text>
          {/*<Text>Blog Posts</Text>*/}
        </View>
        <View style={styles.right}>
          {user && (
            <>
              <Icon
                raised
                reverse
                name="user"
                size={18}
                type="font-awesome"
                color={colorSecondary}
                onPress={onPressProfile}
              />
            </>
          )}
          {!user && (
            <Button
              title="Login"
              onPress={() => {
                // navigation.navigate('Login');
                navigation.navigate('Profile');
              }}
            />
          )}
        </View>
      </View>
      {/*
      <View style={styles.row}>
        <Text>All Posts</Text>
        {user && (
          <>
            <Text>{' | '}</Text>
            <Text>{user.name}'s Posts</Text>
          </>
        )}
      </View>
*/}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    // paddingLeft: 20,
    // paddingRight: 10,
    paddingVertical: 10,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    // alignItems: 'flex-end',
    // justifyContent: 'center',
    // paddingHorizontal: 20,
    // paddingVertical: 10,
  },
  left: {
    flex: 1,
    paddingHorizontal: 20,
    // backgroundColor: 'red',
    // alignContent: 'center',
    // alignSelf: 'center',
    // justifyContent: 'center',
    // textAlignVertical: 'center',
  },
  right: {
    // alignContent: 'center',
    // alignSelf: 'center',
    paddingRight: 10,
  },
});
