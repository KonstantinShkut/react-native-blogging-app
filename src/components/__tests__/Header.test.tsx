import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { Header } from '../Header';
import navigation from '@react-navigation/native';
import { mockStore } from '../../__mocks__/store';

jest.mock('@react-native-firebase/auth', () => jest.fn());
jest.mock('@react-native-firebase/firestore', () => jest.fn());
jest.mock('@react-navigation/native', () => jest.fn());
// @ts-ignore
navigation.useNavigation = jest.fn(() => ({ navigate: jest.fn() }));

describe('Header', () => {
  beforeEach(() => {
    mockStore.mockClear();
  });

  it('renders correctly', () => {
    const component = renderer.create(<Header />);
    expect(component).toMatchSnapshot();
  });
});
