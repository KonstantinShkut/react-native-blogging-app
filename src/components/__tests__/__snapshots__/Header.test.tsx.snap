// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<View
  style={
    Object {
      "paddingLeft": 20,
      "paddingRight": 10,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 33,
            "fontWeight": "bold",
          }
        }
      >
        All Posts
      </Text>
    </View>
    <View>
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "borderRadius": 3,
            "overflow": "hidden",
            "padding": 10,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 0,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2196F3",
                  "fontSize": 18,
                  "paddingBottom": 1,
                  "paddingTop": 2,
                  "textAlign": "center",
                }
              }
            >
              Login
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
